packages/react-instantsearch-dom/src/lib/infiniteHitsCache/sessionStorage.js

```diff
@@ -0,0 +1,49 @@
+import isEqual from 'react-fast-compare';
+
+function getStateWithoutPage(state) {
+  const { page, ...rest } = state;
+  return rest;
+}
+
+const KEY = 'ais.infinitehits';
+
+function hasSessionStorage() {
+  return (
+    typeof window !== 'undefined' &&
+    typeof window.sessionStorage !== 'undefined'
+  );
+}
+
+export default function getInfiniteHitsSessionStorageCache() {
+  return {
+    read({ state }) {
+      if (!hasSessionStorage()) {
+        return null;
+      }
+      try {
+        const cache = JSON.parse(window.sessionStorage.getItem(KEY));
+        return isEqual(cache.state, getStateWithoutPage(state))
+          ? cache.hits
+          : null;
+      } catch (_error) {
+        return null;
```

Comment:
it could also mean that `getItem` throws an error, so deleting from the cache is a good idea, however doing it in a finally or similar could throw an error again, right?

Reply:
Similarly to what Haroen said, there can be an environment where sessionStorage is not available at all like incognito mode of Safari.

Reply:
That can't happen at this stage because [we already check for `sessionStorage` in the lines above](https://github.com/algolia/react-instantsearch/blob/900519a568a3bc0457424d7c34d73f6655853fda/packages/react-instantsearch-dom/src/lib/infiniteHitsCache/sessionStorage.js#L20-L22).

Reply:
Even if it exists, `QuotaExceededError` can be thrown: https://stackoverflow.com/a/18861484

Reply:
Got it! Should we still try to recover from a parsing issue here? Try to delete the cache, otherwise fail silently. Otherwise we might stay with a broken cache "forever".

Reply:
I think JSON parse error won't likely happen unless user directly writes wrong data to sessionStorage with the correct key `ais.infiniteHits`. In our implementation, we write 

```js
window.sessionStorage.setItem(
  KEY,
  JSON.stringify({
    state: getStateWithoutPage(state),
    hits,
  })
)
```

I think it's okay we don't deal with error of low probability that we won't cause.

Reply:
JSON error might still happen if the local storage itself truncates the json for example. In that case we can consider the cache as corrupted and try as if the cache didn't exist

Reply:
Huh, I didn't know sessionStorage truncates data (but not remove). https://github.com/algolia/react-instantsearch/pull/2921/commits/4f191a2b354469c27e9d3288632c09ca72ac45fc